src/samples/iio+network/iio-gyroscope-console-and-mqtt-publish.c

```diff
@@ -0,0 +1,542 @@
+            "\t<sampling frequency>:\t-1=default\n" \
+            "\t<scale>:\t\t<-1=default\n" \
+            "\t<custom offset>:\ty or n\n" \
+            "\t<offset>:\t\tonly take effect if custom offset is \"y\"\n" \
+            "Press CTRL + C to quit\n" \
+            , argv[0]);
+#endif
+        return 0;
+    }
+
+    device_id = sol_iio_address_device(argv[1]);
+    SOL_INT_CHECK_GOTO(device_id, < 0, error_iio);
+
+    SOL_SET_API_VERSION(iio_config.api_version = SOL_IIO_CONFIG_API_VERSION; )
+    iio_config.trigger_name = strdup(argv[2]);
+    SOL_NULL_CHECK(iio_config.trigger_name, -ENOMEM);
+
+    iio_config.buffer_size = atoi(argv[3]);
+    iio_config.sampling_frequency = atoi(argv[4]);
+    channel_config.scale = atof(argv[5]);
+    if (strncmp(argv[6], "y", 1) == 0) {
+        channel_config.use_custom_offset = true;
+        channel_config.offset = atoi(argv[7]);
+    } else
+        channel_config.use_custom_offset = false;
+
+    iio_config.data = &gyro_data;
+    iio_config.sol_iio_reader_cb = iio_gyroscope_reader_cb;
+
+#ifdef MQTT
+    mqtt_config.host = argv[8];
+    mqtt_config.port = atoi(argv[9]);
+    gyro_data.mqtt_topic = argv[10];
+#endif
+
+    device = sol_iio_open(device_id, &iio_config);
+    SOL_NULL_CHECK_GOTO(device, error_iio);
+
+    gyro_data.channel_x = sol_iio_add_channel(device, "in_anglvel_x",
+        &channel_config);
+    SOL_NULL_CHECK_GOTO(gyro_data.channel_x, error_iio);
+
+    gyro_data.channel_y = sol_iio_add_channel(device, "in_anglvel_y",
+        &channel_config);
+    SOL_NULL_CHECK_GOTO(gyro_data.channel_y, error_iio);
+
+    gyro_data.channel_z = sol_iio_add_channel(device, "in_anglvel_z",
+        &channel_config);
+    SOL_NULL_CHECK_GOTO(gyro_data.channel_z, error_iio);
+
+#ifdef GYRO_CALIBRATE
+    reset_calibrate(&(gyro_data.cal_data));
+#endif
+    gyro_data.drop_samples_count = 0;
+
+#ifdef DENOISE_MEDIAN
+    gyro_data.filter_data.buff = (double *)calloc(GYRO_DENOISE_MAX_SAMPLES,
+        sizeof(double) * GYRO_DENOISE_NUM_FIELDS);
+    SOL_NULL_CHECK_GOTO(gyro_data.filter_data.buff, error_iio);
+
+    gyro_data.filter_data.sample_size = GYRO_DENOISE_MAX_SAMPLES;
+    gyro_data.filter_data.count = 0;
+    gyro_data.filter_data.idx = 0;
+#endif
+
+    sol_iio_device_start_buffer(device);
+
+#ifdef MQTT
+    gyro_data.mqtt = sol_mqtt_connect(&mqtt_config);
+    SOL_NULL_CHECK_GOTO(gyro_data.mqtt, error_iio);
+#endif
+
+    sol_run();
+
+    free((char *)iio_config.trigger_name);
+    iio_config.trigger_name = NULL;
+
+#ifdef DENOISE_MEDIAN
+    free(gyro_data.filter_data.buff);
+    gyro_data.filter_data.buff = NULL;
+#endif
+
+    /* Must do sol_iio_close(device), it will disable IIO buffer.
+     * If not, the trigger, buffer size and buffer enable cannot be set
+     * on the next launch.
+     */
+    sol_iio_close(device);
+
+#ifdef MQTT
+    sol_mqtt_disconnect(gyro_data.mqtt);
+#endif
+
+    sol_shutdown();
+    return 0;
+
+error_iio:
```

Comment:
you should call `sol_shutdown()` in this label too because `sol_init()` was already called.